[Core] Fix UTF-8 support in positive character classes

vasia123 opened the pull request:
Fixes #138: Positive character classes with non-ASCII characters (e.g., [а-я], [一-龥], [😀-😿]) were not matching correctly, while negative classes worked fine.

Root cause: In CharacterClass(), Unicode codepoints were truncated to uint8_t before adding FSM edges, losing the upper bytes.

Changes:
- Add CodepointToPackedUTF8() helper to convert Unicode codepoints to packed UTF-8 format for FSM edge construction
- Update CharacterClass() to use proper UTF-8 encoding for ranges
- Add UTF-8 support in AdvanceCharacterClass() and AdvanceCharacterClassStar() for non-FSM fallback path
- Add partial_codepoint field to ParserState for UTF-8 accumulation

Tests added for:
- Cyrillic characters (2-byte UTF-8)
- CJK characters (3-byte UTF-8)
- Emoji (4-byte UTF-8)
- Mixed UTF-8 ranges
- Single non-ASCII character classes

Seven-Streams commented on tests/python/test_grammar_matcher_ebnf.py:
```diff
@@ -500,6 +500,144 @@ def test_character_class_star_utf8():
+    assert not _is_grammar_accept_string(grammar, "")  # empty
+
+
+def test_positive_utf8_character_class_mixed_ranges():
```
Please also add a test for the ebnf grammar:

```
"root ::= [a-zа-я一-龥]{0, 2048}"
```

Seven-Streams commented:
I'll check it.

vasia123 commented:
@Seven-Streams 

Thank you for the review! I've addressed the feedback:

1. Updated `kXGrammarSerializeVersion` from v7 to v8
2. Updated version references in `test_serialization.py`
3. Added test `test_positive_utf8_character_class_with_quantifier` for `[a-zа-я一-龥]{0,2048}`
4. Added early rejection check for partial UTF-8 sequences in both `AdvanceCharacterClass` and `AdvanceCharacterClassStar` when `sub_element_id != 0`

Regarding the mask comparison test - could you clarify what you'd like compared? The existing UTF-8 tests verify correct accept/reject behavior. Did you mean a test similar to `test_fill_next_token_bitmask` that checks specific rejected token counts?

Seven-Streams commented:
@vasia123 Exactly. I'm wondering if you can add a test for Unicode characters similar to `test_fill_next_token_bitmask`, to check the rejected token counts. But considering that you contribute a new feature, maybe we don't need to add one. @Ubospica cc.
